applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,743 @@
+	u16_t eval_success_threshold;
+} __packed;
+
+struct params_wifi {
+	s16_t wifi_rating_inc;
+	s16_t wifi_present_threshold;
+	s16_t wifi_active_threshold;
+} __packed;
+
+struct params_chmap {
+	u8_t chmap[CHMAP_BLE_BITMASK_SIZE];
+} __packed;
+
+struct params_blacklist {
+	u16_t wifi_chn_bitmask;
+} __packed;
+
+static u8_t chmap_instance_buf[CHMAP_FILTER_INST_SIZE];
+static struct chmap_instance *chmap_inst;
+static u8_t current_chmap[CHMAP_BLE_BITMASK_SIZE] = CHMAP_BLE_BITMASK_DEFAULT;
+static atomic_t processing;
+static atomic_t new_blacklist;
+static atomic_t params_updated;
+static struct chmap_filter_params filter_params;
+static struct k_mutex data_access_mutex;
+
+BUILD_ASSERT(sizeof(struct bt_hci_cp_le_set_host_chan_classif) ==
+	     sizeof(struct params_chmap));
+BUILD_ASSERT(sizeof(current_chmap) == sizeof(struct params_chmap));
+BUILD_ASSERT(THREAD_PRIORITY >= CONFIG_BT_HCI_TX_PRIO);
+
+static void ble_qos_thread_fn(void);
+
+static struct device *cdc_dev;
+static u32_t cdc_dtr;
+
+static void ble_chn_stats_print(bool update_channel_map)
+{
+	char str[64];
+	int str_len;
+	s16_t chn_rating;
+	u8_t chn_freq, chn_state;
+
+	if (!IS_ENABLED(CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE)) {
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_UART_LINE_CTRL)) {
+		int err;
+		u32_t cdc_val;
+
+		/* Repeated to monitor CDC state */
+		err = uart_line_ctrl_get(cdc_dev, LINE_CTRL_DTR, &cdc_val);
+		if (!err) {
+			cdc_dtr = cdc_val;
+		}
+
+		if (!cdc_dtr) {
+			return;
+		}
+	}
+
+	/* Note: if UART FIFO overflows, expect a warning printout: */
+	/* "<wrn> usb_cdc_acm: Ring buffer full (...)" */
+	/* USB_CDC_ACM_RINGBUF_SIZE can be increased to prevent an overflow */
+
+	if (update_channel_map) {
+		str_len = snprintf(
+			str,
+			sizeof(str),
+			"[%08" PRIu32 "]Channel map update\n",
+			(u32_t)k_cycle_get_32());
+		if (str_len <= 0 || str_len > sizeof(str)) {
+			LOG_ERR("Encoding error");
+			return;
+		}
+		uart_fifo_fill(cdc_dev, str, str_len);
+	}
+
+	/* Channel state information print format: */
+	/* BT_INFO={idx0_freq:idx0_state:idx0_rating, ... } */
+
+	str_len = snprintf(str, sizeof(str), "BT_INFO={");
+	if (str_len <= 0 || str_len > sizeof(str)) {
+		LOG_ERR("Encoding error");
+		return;
+	}
+	uart_fifo_fill(cdc_dev, str, str_len);
+
+	str_len = 0;
+	for (u8_t i = 0; i < CHMAP_BLE_CHANNEL_COUNT; i++) {
+		int part_len;
+
+		chmap_filter_chn_info_get(
+			chmap_inst,
+			i,
+			&chn_state,
+			&chn_rating,
+			&chn_freq);
+		part_len = snprintf(
+			&str[str_len],
+			sizeof(str) - str_len,
+			"%" PRIu32 ":%" PRIu32 ":%" PRId16 ",",
+			chn_freq,
+			chn_state,
+			chn_rating);
+		str_len += part_len;
+		if (part_len <= 0 || str_len > sizeof(str)) {
+			LOG_ERR("Encoding error");
+			return;
+		}
+
+		if (str_len >= ((sizeof(str) * 2) / 3)) {
+			uart_fifo_fill(cdc_dev, str, str_len);
+			str_len = 0;
+		}
+	}
+	str_len += snprintf(&str[str_len], sizeof(str) - str_len, "}\n");
+	if (str_len <= 0 || str_len > sizeof(str)) {
```

Comment:
at this point str_len is large positive value so error will not be handled as intended
```
str_len += snprintf(...)
`str_len <= 0`
```

Reply:
Missed that one. Adding temp variable to hold snprintf return